File: gobron-backend/app/models/manual_booking.py
"""Manual bookings created by field owners before online booking is primary."""
from datetime import date, datetime, time
from decimal import Decimal

from sqlalchemy import Date, DateTime, Enum, ForeignKey, Numeric, String, Text, Time, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.enums import ManualBookingStatus


class ManualBooking(Base):
    __tablename__ = "manual_bookings"

    id: Mapped[int] = mapped_column(primary_key=True)
    owner_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    field_id: Mapped[int] = mapped_column(
        ForeignKey("fields.id", ondelete="CASCADE"), index=True
    )

    booking_date: Mapped[date] = mapped_column(Date, index=True)
    start_time: Mapped[time] = mapped_column(Time, nullable=False)
    end_time: Mapped[time] = mapped_column(Time, nullable=False)
    customer_name: Mapped[str | None] = mapped_column(String(120))
    customer_phone: Mapped[str | None] = mapped_column(String(20))
    price: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    note: Mapped[str | None] = mapped_column(Text)
    status: Mapped[ManualBookingStatus] = mapped_column(
        Enum(
            ManualBookingStatus,
            name="manual_booking_status",
            values_callable=lambda enum_cls: [member.value for member in enum_cls],
        ),
        default=ManualBookingStatus.BOOKED,
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    field: Mapped["Field"] = relationship()  # noqa: F821

File: gobron-flutter-admin/lib/features/manual_bookings/models/manual_booking.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

import '../../fields/models/field.dart';

enum ManualBookingStatus { booked, cancelled, completed }

class ManualBooking {
  ManualBooking({
    required this.id,
    required this.ownerId,
    required this.fieldId,
    required this.bookingDate,
    required this.startTime,
    required this.endTime,
    this.customerName,
    this.customerPhone,
    required this.price,
    this.note,
    required this.status,
  });

  factory ManualBooking.fromJson(Map<String, dynamic> json) {
    return ManualBooking(
      id: json['id'] as int,
      ownerId: json['owner_id'] as int,
      fieldId: json['field_id'] as int,
      bookingDate: DateTime.parse(json['booking_date'] as String),
      startTime: Field.timeFromJson(json['start_time'] as String),
      endTime: Field.timeFromJson(json['end_time'] as String),
      customerName: json['customer_name'] as String?,
      customerPhone: json['customer_phone'] as String?,
      price: _toDouble(json['price']),
      note: json['note'] as String?,
      status: ManualBookingStatus.values.byName(json['status'] as String),
    );
  }

  final int id;
  final int ownerId;
  final int fieldId;
  final DateTime bookingDate;
  final TimeOfDay startTime;
  final TimeOfDay endTime;
  final String? customerName;
  final String? customerPhone;
  final double price;
  final String? note;
  final ManualBookingStatus status;

  static double _toDouble(dynamic v) =>
      v is String ? double.parse(v) : (v as num).toDouble();
}

class ManualBookingInput {
  ManualBookingInput({
    required this.fieldId,
    required this.bookingDate,
    required this.startTime,
    required this.endTime,
    this.customerName,
    this.customerPhone,
    required this.price,
    this.note,
  });

  Map<String, dynamic> toJson() => {
    'field_id': fieldId,
    'booking_date': DateFormat('yyyy-MM-dd').format(bookingDate),
    'start_time': Field.timeToJson(startTime),
    'end_time': Field.timeToJson(endTime),
    'customer_name': customerName,
    'customer_phone': customerPhone,
    'price': price.toStringAsFixed(2),
    'note': note,
  };

  final int fieldId;
  final DateTime bookingDate;
  final TimeOfDay startTime;
  final TimeOfDay endTime;
  final String? customerName;
  final String? customerPhone;
  final double price;
  final String? note;
}
